syntax = "proto3";

package product;

option go_package = "./product";

// 商品服务定义
service ProductService {
  // 查询单个商品信息
  rpc GetProduct(GetProductRequest) returns (GetProductResponse) {}
  
  // 批量查询商品信息
  rpc GetProducts(GetProductsRequest) returns (GetProductsResponse) {}
  
  // 创建商品（可选）
  rpc CreateProduct(CreateProductRequest) returns (CreateProductResponse) {}
  
  // 更新商品信息（可选）
  rpc UpdateProduct(UpdateProductRequest) returns (UpdateProductResponse) {}
  
  // 删除商品（可选）
  rpc DeleteProduct(DeleteProductRequest) returns (DeleteProductResponse) {}
  
  // 减少库存
  rpc ReduceStock(ReduceStockRequest) returns (ReduceStockResponse) {}
}

// 商品信息
message Product {
  int32 id = 1;
  string name = 2;
  string description = 3;
  double price = 4;
  int32 stock = 5;
  string category = 6;
  string image_url = 7;
  string created_at = 8;
  string updated_at = 9;
}

// 查询单个商品请求
message GetProductRequest {
  int32 product_id = 1;
}

// 查询单个商品响应
message GetProductResponse {
  Product product = 1;
  bool success = 2;
  string error_message = 3;
}

// 批量查询商品请求
message GetProductsRequest {
  repeated int32 product_ids = 1;  // 指定ID列表查询
  string category = 2;             // 按分类查询
  int32 page = 3;                  // 分页参数
  int32 page_size = 4;             // 每页数量
}

// 批量查询商品响应
message GetProductsResponse {
  repeated Product products = 1;
  int32 total = 2;                 // 总商品数
  bool success = 3;
  string error_message = 4;
}

// 创建商品请求
message CreateProductRequest {
  string name = 1;
  string description = 2;
  double price = 3;
  int32 stock = 4;
  string category = 5;
  string image_url = 6;
}

// 创建商品响应
message CreateProductResponse {
  int32 product_id = 1;
  bool success = 2;
  string error_message = 3;
}

// 更新商品请求
message UpdateProductRequest {
  int32 product_id = 1;
  string name = 2;
  string description = 3;
  double price = 4;
  int32 stock = 5;
  string category = 6;
  string image_url = 7;
}

// 更新商品响应
message UpdateProductResponse {
  bool success = 1;
  string error_message = 2;
}

// 删除商品请求
message DeleteProductRequest {
  int32 product_id = 1;
}

// 删除商品响应
message DeleteProductResponse {
  bool success = 1;
  string error_message = 2;
}

// 减少库存请求
message ReduceStockRequest {
  int32 product_id = 1;
  int32 quantity = 2;
}

// 减少库存响应
message ReduceStockResponse {
  bool success = 1;
  string error_message = 2;
}